src/runtime-rs/crates/resource/src/cpu_mem/mem.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{Context, Ok, Result};
+use hypervisor::Hypervisor;
+use kata_types::config::TomlConfig;
+use oci::LinuxResources;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+use crate::ResourceUpdateOp;
+
+// MIBTOBYTESSHIFT the number to shift needed to convert MiB to Bytes
+pub const MIB_TO_BYTES_SHIFT: i32 = 20;
+#[derive(Default, Debug, Clone)]
+pub struct MemResource {
+    /// Current memory
+    pub(crate) current_mem: Arc<RwLock<u32>>,
+
+    /// Default memory
+    pub(crate) default_mem: u32,
+
+    /// MemResource of each container
+    pub(crate) container_mem_resources: Arc<RwLock<HashMap<String, LinuxResources>>>,
+
+    /// Use guest swap
+    pub(crate) use_guest_swap: bool,
+}
+
+impl MemResource {
+    pub fn new(config: Arc<TomlConfig>) -> Result<Self> {
+        let hypervisor_name = config.runtime.hypervisor_name.clone();
+        let hypervisor_config = config
+            .hypervisor
+            .get(&hypervisor_name)
+            .context("failed to get hypervisor")?;
+
+        Ok(Self {
+            current_mem: Arc::new(RwLock::new(hypervisor_config.memory_info.default_memory)),
+            container_mem_resources: Arc::new(RwLock::new(HashMap::new())),
+            use_guest_swap: hypervisor_config.memory_info.enable_guest_swap,
+            default_mem: hypervisor_config.memory_info.default_memory,
+        })
+    }
+
+    pub(crate) async fn update_mem_resources(
+        &self,
+        cid: &str,
+        linux_resources: Option<&LinuxResources>,
+        op: ResourceUpdateOp,
+        hypervisor: &dyn Hypervisor,
+    ) -> Result<()> {
+        self.update_container_mem_resources(cid, linux_resources, op)
+            .await
+            .context("update container cpu resources")?;
+        // the unit here is byte
+        let (mem_sb_byte, need_pod_swap, swap_sb_byte) = self
+            .total_mems(self.use_guest_swap)
+            .await
+            .context("failed to calculate total memory requirement for containers")?;
+        let mut mem_sb_mb = (mem_sb_byte >> MIB_TO_BYTES_SHIFT) as u32;
+        mem_sb_mb += self.default_mem;
```

Comment:
I am not sure if just adding the `default_mem` is proper here.

If annotations of initial size are passed from the upper layer runtime,  we will modify the `default_mem`. So the `default_mem` will no longer be the one we set in the configuration file.

Reply:
fixed , I store the old one in `MemResource` now